File: app/src/main/java/org/ngbp/jsonrpc4jtestharness/http/servers/MiddlewareWebServer.java
package org.ngbp.jsonrpc4jtestharness.http.servers;

import android.content.Context;

import org.eclipse.jetty.http.HttpVersion;
import org.eclipse.jetty.server.Connector;
import org.eclipse.jetty.server.Handler;
import org.eclipse.jetty.server.HttpConfiguration;
import org.eclipse.jetty.server.HttpConnectionFactory;
import org.eclipse.jetty.server.SecureRequestCustomizer;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.ServerConnector;
import org.eclipse.jetty.server.SslConnectionFactory;
import org.eclipse.jetty.server.handler.ContextHandler;
import org.eclipse.jetty.server.handler.HandlerList;
import org.eclipse.jetty.servlet.ServletContextHandler;
import org.eclipse.jetty.servlet.ServletHolder;
import org.eclipse.jetty.util.ssl.SslContextFactory;
import org.eclipse.jetty.websocket.server.WebSocketHandler;
import org.eclipse.jetty.websocket.servlet.WebSocketServletFactory;
import org.ngbp.jsonrpc4jtestharness.R;
import org.ngbp.jsonrpc4jtestharness.core.ws.MiddlewareWebSocket;

import java.io.InputStream;
import java.security.KeyStore;

import javax.net.ssl.KeyManagerFactory;
import javax.net.ssl.SSLContext;
import javax.net.ssl.TrustManagerFactory;

public class MiddlewareWebServer implements AutoCloseable {

    public MiddlewareWebServer(Context context) {
        this.context = context;
    }

    public static class MiddlewareSocketHandler extends WebSocketHandler {
        @Override
        public void configure(WebSocketServletFactory factory)
        {
            factory.register(MiddlewareWebSocket.class);
        }
    }

    private Server server;
    private Context context;

    public void runWebServer() throws Exception {

        server = new Server();

        InputStream i = context.getResources().openRawResource(R.raw.mykey);

        KeyStore keystore = KeyStore.getInstance("PKCS12");
        try {
            keystore.load(i, "MY_PASSWORD".toCharArray());
        } finally {
            i.close();
        }

        KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance("X509");
        keyManagerFactory.init(keystore, "MY_PASSWORD".toCharArray());
        TrustManagerFactory tmf = TrustManagerFactory.getInstance("X509");
        tmf.init(keystore);

        SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(keyManagerFactory.getKeyManagers(), tmf.getTrustManagers(), null);

        // HTTP connector
        ServerConnector connector = new ServerConnector(server);
        connector.setPort(8081);

        // HTTPS configuration
        HttpConfiguration https = new HttpConfiguration();
        https.addCustomizer(new SecureRequestCustomizer());
        https.setSecurePort(8443);

        HttpConfiguration wss = new HttpConfiguration();
        https.addCustomizer(new SecureRequestCustomizer());
        https.setSecurePort(9999);

        // Configuring SSL
        SslContextFactory sslContextFactory = new SslContextFactory.Server();
        sslContextFactory.setKeyStoreType("PKCS12");

        // Defining keystore path and passwords
        sslContextFactory.setSslContext(sslContext);
        sslContextFactory.setKeyStorePassword("MY_PASSWORD");
        sslContextFactory.setKeyManagerPassword("MY_PASSWORD");

        // Configuring the connector
        ServerConnector sslConnector = new ServerConnector(server, new SslConnectionFactory(sslContextFactory, HttpVersion.HTTP_1_1.toString()), new HttpConnectionFactory(https));
        sslConnector.setPort(8443);

        ServerConnector wssConnector = new ServerConnector(server, new SslConnectionFactory(sslContextFactory, HttpVersion.HTTP_1_1.toString()), new HttpConnectionFactory(wss));
        sslConnector.setPort(9999);

        // Setting HTTP and HTTPS connectors
        server.setConnectors(new Connector[]{connector, sslConnector, wssConnector});

        ServletContextHandler handler = new ServletContextHandler(server, "/");
        JsonRpcTestServlet servlet = new JsonRpcTestServlet(context);
        ServletHolder holder = new ServletHolder(servlet);
        handler.addServlet(holder, "/github");

        ContextHandler ch = new ContextHandler(server,"/echo");
        ch.setHandler(new MiddlewareSocketHandler());

        HandlerList handlers = new HandlerList();
        handlers.setHandlers(new Handler[] {new MiddlewareSocketHandler(), handler}); //Added socket handler

        server.setHandler(handlers);

        try {
            server.start();
            server.join();
        } catch (Throwable t) {
            t.printStackTrace(System.err);
        }
    }

    @Override
    public void close() throws Exception {
        this.stop();
    }

    public void stop() throws Exception {
        server.stop();
    }

    public Server getServer() {
        return server;
    }

    public void setServer(Server server) {
        this.server = server;
    }
}

File: app/src/main/java/org/ngbp/jsonrpc4jtestharness/http/servers/MiddlewareWebServer.kt
package org.ngbp.jsonrpc4jtestharness.http.servers

import org.eclipse.jetty.http.HttpVersion
import org.eclipse.jetty.server.*
import org.eclipse.jetty.server.handler.HandlerList
import org.eclipse.jetty.server.handler.ResourceHandler
import org.eclipse.jetty.util.resource.Resource
import org.eclipse.jetty.util.ssl.SslContextFactory
import org.eclipse.jetty.websocket.server.WebSocketHandler
import org.eclipse.jetty.websocket.servlet.ServletUpgradeRequest
import org.eclipse.jetty.websocket.servlet.ServletUpgradeResponse
import org.eclipse.jetty.websocket.servlet.WebSocketCreator
import org.eclipse.jetty.websocket.servlet.WebSocketServletFactory
import org.ngbp.jsonrpc4jtestharness.core.ws.IUserAgentSSLContext
import org.ngbp.jsonrpc4jtestharness.core.ws.MiddlewareWebSocket
import org.ngbp.jsonrpc4jtestharness.rpc.processor.RPCProcessor
import java.io.IOException
import java.security.GeneralSecurityException
import java.util.*


class MiddlewareWebServer private constructor(builder: Builder) : AutoCloseable {

    private val httpsPort: Int
    private val httpPort: Int
    private val wssPort: Int
    private val wsPort: Int
    private val hostName: String?
    private val rpcProcessor: RPCProcessor?
    private val connectors: Array<Connectors>
    private val generatedSSLContext: IUserAgentSSLContext?

    enum class Connectors {
        HTTP_CONNECTOR, HTTPS_CONNECTOR, WS_CONNECTOR, WSS_CONNECTOR
    }

    companion object {
        const val HTTP_PORT = 8080
        const val HTTPS_PORT = 8443
        const val WSS_PORT = 9999
        const val WS_PORT = 9998
    }

    class Builder {
        var httpsPort: Int = HTTPS_PORT
        var httpPort: Int = HTTP_PORT
        var wssPort: Int = WSS_PORT
        var wsPort: Int = WS_PORT
        var hostName: String? = null
        var rpcProcessor: RPCProcessor? = null
        var connectors: Array<Connectors> = arrayOf()
        var generatedSSLContext: IUserAgentSSLContext? = null

        fun httpsPort(value : Int) = apply { httpsPort = value }

        fun httpPort(value : Int) = apply { httpPort = value }

        fun wssPort(value : Int) = apply { wssPort = value }

        fun wsPort(value : Int) = apply { wsPort = value }

        fun hostName(value : String?) = apply { hostName = value }

        fun addRPCProcessor(value: RPCProcessor) = apply { rpcProcessor = value }

        fun enableConnectors(value : Array<Connectors>) = apply { connectors = value }

        fun sslContext(value : IUserAgentSSLContext?) = apply { generatedSSLContext = value }

        fun build() = MiddlewareWebServer(this)
    }

    private var server: Server = Server()
        private set(value) {
            if (!value.isRunning) field = value
        }

    private var sslContextFactory: SslContextFactory? = null

    @Throws(Exception::class)
    private fun prepareServer() {
        configureSSLFactory()
        configureConnectors()
        configureHandlers()
    }

    @Throws(GeneralSecurityException::class, IOException::class)
    private fun configureSSLFactory() {
        if (generatedSSLContext != null) {
            // Configuring SSL
            sslContextFactory = SslContextFactory.Server().apply {
                keyStoreType = "PKCS12"
                sslContext = generatedSSLContext.getInitializedSSLContext("MY_PASSWORD")
                setKeyStorePassword("MY_PASSWORD")
                setKeyManagerPassword("MY_PASSWORD")
            }
        }
    }

    private fun configureConnectors() {
        val enabledConnectors: MutableList<Connector> = ArrayList()

        // HTTP connector
        if (connectors.contains(Connectors.HTTP_CONNECTOR)) {
            val httpConnector = getServerConnector(httpPort)
            enabledConnectors.add(httpConnector)
        }

        // WS configuration
        if (connectors.contains(Connectors.WS_CONNECTOR)) {
            val wsConnector = getServerConnector(wsPort)
            enabledConnectors.add(wsConnector)
        }
        if (sslContextFactory != null) {
            // HTTPS configuration
            if (connectors.contains(Connectors.HTTPS_CONNECTOR)) {
                val httpsConnector = getSecureServerConnector(sslContextFactory, httpsPort)
                enabledConnectors.add(httpsConnector)
            }

            // WSS configuration
            if (connectors.contains(Connectors.WSS_CONNECTOR)) {
                val wssConnector = getSecureServerConnector(sslContextFactory, wssPort)
                enabledConnectors.add(wssConnector)
            }
        }

        // Setting HTTP, HTTPS, WS and WSS connectors
        if (enabledConnectors.isNotEmpty()) {
            server.connectors = enabledConnectors.toTypedArray()
        }
    }

    inner class MiddlewareWebSocketCreator : WebSocketCreator {
        override fun createWebSocket(req: ServletUpgradeRequest, resp: ServletUpgradeResponse): Any? {
            return rpcProcessor?.let { MiddlewareWebSocket(it) }
        }
    }

    private fun configureHandlers() {
        val handlerArray: MutableList<Handler> = ArrayList()

        val resourceHandler = ResourceHandler()
        resourceHandler.isDirectoriesListed = true
        resourceHandler.welcomeFiles = arrayOf("index.html")
        resourceHandler.baseResource = Resource.newResource("storage/emulated/0/Download/test") // /route/5003/app/
        handlerArray.add(resourceHandler)

        if (rpcProcessor != null) {
            val webSocketHandler: WebSocketHandler = object : WebSocketHandler() {
                override fun configure(factory: WebSocketServletFactory) {
                    factory.creator = MiddlewareWebSocketCreator()
                }
            }
            handlerArray.add(webSocketHandler)
        }

        if (handlerArray.isNotEmpty()) {
            val handlers = HandlerList()
            handlers.handlers = handlerArray.toTypedArray()
            server.handler = handlers
        }
    }

    private fun getServerConnector(port: Int): ServerConnector {
        val connector = ServerConnector(server)
        connector.port = port

        if (hostName != null) {
            if (hostName.isEmpty()) {
                connector.host = hostName
            }
        }
        return connector
    }

    private fun getSecureServerConnector(sslContextFactory: SslContextFactory?, securePort: Int): ServerConnector {
        val config = HttpConfiguration()
        config.addCustomizer(SecureRequestCustomizer())
        config.securePort = securePort

        // Configuring the secure connector
        val connector = ServerConnector(server, SslConnectionFactory(sslContextFactory, HttpVersion.HTTP_1_1.toString()), HttpConnectionFactory(config))
        connector.port = securePort
        if (hostName != null) {
            if (hostName.isEmpty()) {
                connector.host = hostName
            }
        }
        return connector
    }

    fun isRunning() = server.isRunning

    @Throws(MiddlewareWebServerError::class)
    fun start() {
        server.start()
    }

    @Throws(Exception::class)
    fun stop() {
        server.stop()
    }

    @Throws(Exception::class)
    override fun close() {
        stop()
    }

    init {
        httpsPort = builder.httpsPort
        httpPort = builder.httpPort
        wssPort = builder.wssPort
        wsPort = builder.wsPort
        hostName = builder.hostName
        rpcProcessor = builder.rpcProcessor
        connectors = builder.connectors
        generatedSSLContext = builder.generatedSSLContext

        prepareServer()
    }
}